pyop2/array.py

```diff
@@ -0,0 +1,171 @@
+        elif configuration["backend"] == "OPENCL":
+            return OpenCLArray(*args, **kwargs)
+        else:
+            raise ValueError
+
+    @property
+    def kernel_arg(self):
+        # lazy but for now assume that the data is always modified if we access
+        # the pointer
+        return self.device_ptr if offloading else self.host_ptr
+
+        # if offloading:
+        #     # N.B. not considering MAX, MIN here
+        #     if access in {RW, WRITE, INC}:
+        #         return self.device_ptr
+        #     else:
+        #         return self.device_ptr_ro
+        # else:
+        #     # N.B. not considering MAX, MIN here
+        #     if access in {RW, WRITE, INC}:
+        #         return self.host_ptr
+        #     else:
+        #         return self.host_ptr_ro
+
+    @property
+    def is_available_on_device(self):
+        return self.availability in {ON_DEVICE, ON_BOTH}
+
+    def ensure_availability_on_device(self):
+        if not self.is_available_on_device:
+            self.host_to_device_copy()
+
+    @property
+    def is_available_on_host(self):
+        return self.availability in {ON_HOST, ON_BOTH}
+
+    def ensure_availability_on_host(self):
+        if not self.is_available_on_host:
+            self.device_to_host_copy()
+
+    @property
+    def vec(self):
+        raise NotImplementedError
+        # if offload == "cpu":
+        #     return PETSc.Vec().createMPI(...)
+        # elif offload == "gpu":
+        #     if backend == "cuda":
+        #         return PETSc.Vec().createCUDA(..)
+        #     ...
+
+
+    @property
+    def data(self):
+        self.ensure_availability_on_host()
+        self.availability = ON_HOST
+        v = self._host_data.view()
+        v.setflags(write=True)
+        return v
+
+    @property
+    def data_ro(self):
+        self.ensure_availability_on_host()
+        v = self._host_data.view()
+        v.setflags(write=False)
+        return v
+
+    @property
+    def host_ptr(self):
+        self.ensure_availability_on_host()
+        self.availability = ON_HOST
+        return self.data.ctypes.data
+
+    @property
+    def host_ptr_ro(self):
+        self.ensure_availability_on_host()
+        return self.data.ctypes.data
+
+
+class CPUOnlyArray(MirroredArray):
+
+    @property
+    def device_ptr(self):
+        return self.host_ptr
+
+    @property
+    def device_ptr_ro(self):
+        return self.host_ptr_ro
+
+    def host_to_device_copy(self):
+        pass
+
+    def device_to_host_copy(self):
+        pass
+
+
+if configuration["backend"] == "OPENCL":
+    # TODO: Instruct the user to pass
+    # -viennacl_backend opencl
+    # -viennacl_opencl_device_type gpu
+    # create a dummy vector and extract its associated command queue
+    x = PETSc.Vec().create(PETSc.COMM_WORLD)
+    x.setType("viennacl")
+    x.setSizes(size=1)
+    queue_ptr = x.getCLQueueHandle()
+    cl_queue = pyopencl.CommandQueue.from_int_ptr(queue_ptr, retain=False)
```

Comment:
Possibly. What I'm trying to illustrate here is that we can do without the `OpenCLBackend` class since we don't need to maintain different `Dat`, `Global`, etc subclasses.